### PR TITLE
Fixing the the mip gpu copy for compressed format texture in GL41Backend

### PR DESCRIPTION
**Pull request by samcake**
This PR fixes a new bug that we encountered by introducing the support for compressed texture in baked content.

On GL41Backend:
 Macos or 
 Windows on non 4.5 capable gpu
                 or forced with HIFI_DISABLE_OPENGL_45 env var defined

The problem is that texture with compressed format cannot be attached to FBO and so that technique (used for non compressed textures) fails. As the texture gets promoted (or demoted), a new gltexture is created and the already existing mip levels are copied directly in gpu in the new texture.
This copy was failing for compressed textures and so some mips are showing BLACK or garbage as a texture gets promoted and these mip levels are seen.
This is not happening with this pr.

## TEST PLAN
Testing on macos or windows with supported GPU but with the var env "HIFI_DISABLE_OPENGL_45" defined.
With a cold ktx cache (erased in appData/Local/Highfidelity/Interface) go to  dev-welcome and load the scene. go close and then far from any of the grass patches.
In current master, this should start to show the bug with some texture turning black in the distance.
Not happening with this pr.

Another way to test this pr is to change the Texture Memory Budget from small to large (in the Developer/Render/Maximum Texture Memory menu)
And check that the textures are gaining resolution or loosing resolution but don;t get corrupted.
In current master, we see corruption.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: d7dfab6405c4362b13a1b3bd743461ee4942b743 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-d7dfab6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-d7dfab6.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8240c3f76112d9c1b300eb50211886dda732ef3e available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-8240c3f.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-8240c3f.dmg

File: libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp [hunk 1]
```diff
@@ -362,35 +445,115 @@ void GL41VariableAllocationTexture::demote() {
     auto oldSize = _size;
     const_cast<GLuint&>(_id) = allocate(_gpuObject);
     uint16_t oldAllocatedMip = _allocatedMip;
+    uint16_t oldPopulatedMip = _populatedMip;
```
Comment (jherico) [inline libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp]: this logic doesn't make sense to me.  if the _populatedMip was 0 and the new allocated mip is 1, then why would you copy the old mip 0 to the PBO?  You're not going to have anywhere to put it in the new texture.

Comment (samcake) [inline libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp]: good point looking at it

**Comment by samcake** [comment]
Still having issues and not the proper behavior, exploring...

**Comment by hifi-gustavo** [comment]
Build console output for commit: d0fa8fe88713ae4be7ad3e7662280cffca788758 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-d0fa8fe.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-d0fa8fe.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 83618d45a23ec97d4b78502a0f06549d1e139f93 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-83618d4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-83618d4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bfe8bc7593b0710864a041333aecc8757de1550c available at: https://gist.github.com/acfe865630d7fecbfb42876a0cdb5042<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-bfe8bc7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-bfe8bc7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 04895864b5f9c3fc543ac71584fe3f393b2605a4 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-0489586.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-0489586.dmg

File: libraries/gpu/src/gpu/Texture.h [hunk 2]
```diff
@@ -372,7 +372,7 @@ class Texture : public Resource {
     bool isColorRenderTarget() const;
     bool isDepthStencilRenderTarget() const;
 
-    const Element& getTexelFormat() const { return _texelFormat; }
+    const Element getTexelFormat() const { return _texelFormat; }
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.h]: I don't get why you're removing the reference here.  it doesn't make the code any more thread safe.  If you're removing the reference though, you should remove the const.  const on a returned type doesn't make any sense and if you do 
`Element element = tex->getTexelFormat()` then having const just ends up triggering another copy.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ae5cd3059ab16d4b8c16a73742312a44b18b88c4 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-ae5cd30.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-ae5cd30.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3abd9f21e80c35f5185b4a830e03b679a7d6a798 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-3abd9f2.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-3abd9f2.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8b62fad85eaadd7b1a96b859d03157e447ddc9fd available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-8b62fad.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-8b62fad.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 54ace01cd6f962f1b43faed541e2f24510cf470d available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-54ace01.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-54ace01.dmg

File: libraries/gpu/src/gpu/Texture.h (hunk 2 above)
Comment (samcake) [inline libraries/gpu/src/gpu/Texture.h]: The element is a 2 bytes thing, it s simpler to just copy it
I ll remove the const, good point

**Comment by hifi-gustavo** [comment]
Build console output for commit: e561a33b0165d51daeeb53c2e9643778dcb9c8fd available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-e561a33.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-e561a33.dmg

**Comment by samcake** [comment]
Ready to QA / CR again

**Comment by hifi-gustavo** [comment]
Build console output for commit: 32d675c82bee798b8d8e25ed9c98c19ddd2d19d1 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-32d675c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-32d675c.dmg

**Comment by marko8904** [comment]
test this please

File: libraries/gpu-gl/src/gpu/gl41/GL41Backend.h [hunk 3]
```diff
@@ -37,12 +37,16 @@ class GL41Backend : public GLBackend {
 public:
     static const GLint TRANSFORM_OBJECT_SLOT  { 31 };
     static const GLint RESOURCE_TRANSFER_TEX_UNIT { 32 };
-    static const GLint RESOURCE_BUFFER_TEXBUF_TEX_UNIT { 33 };
-    static const GLint RESOURCE_BUFFER_SLOT0_TEX_UNIT { 34 };
+    static const GLint RESOURCE_TRANSFER_EXTRA_TEX_UNIT { 33 };
```
Comment (jherico) [inline libraries/gpu-gl/src/gpu/gl41/GL41Backend.h]: are we sure all our target hardware (particularly on mac) supports more than 32 texture units?

File: libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp [hunk 4]
```diff
@@ -306,6 +308,130 @@ void GL41VariableAllocationTexture::syncSampler() const {
...
+            (void)CHECK_GL_ERROR();
+        }
+    }
+
+    // destroy the transfer framebuffer
+    glBindFramebuffer(GL_READ_FRAMEBUFFER, 0);
+    glDeleteFramebuffers(1, &fbo);
+}
+
+void copyCompressedTexGPUMem(const gpu::Texture& texture, GLenum texTarget, GLuint srcId, GLuint destId, uint16_t numMips, uint16_t srcAllocatedMip, uint16_t destAllocatedMip, uint16_t populatedMips) {
```
Comment (jherico) [inline libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp]: should probably be `srcMipOffset` and `destMipOffset` for clarity, since this isn't a member function and doesn't have the concept of `_allocatedMip` being relative to all the mips described in the `gpu::Texture` object.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 32d675c82bee798b8d8e25ed9c98c19ddd2d19d1 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-32d675c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-32d675c.dmg

File: libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp [hunk 5]
```diff
@@ -360,37 +475,26 @@ void GL41VariableAllocationTexture::demote() {
...
-        // destroy the transfer framebuffer
-        glBindFramebuffer(GL_READ_FRAMEBUFFER, 0);
-        glDeleteFramebuffers(1, &fbo);
-
-        syncSampler();
-    });
-
+    // copy pre-existing mips
+    uint16_t numMips = _gpuObject.getNumMips();
+    if (_texelFormat.isCompressed()) {
```
Comment (jherico) [inline libraries/gpu-gl/src/gpu/gl41/GL41BackendTexture.cpp]: Could be 

````
withPreservedTexture([&] {
    if (_texelFormat.isCompressed()) {
        copyCompressedTexGPUMem(_gpuObject, _target, oldId, _id, numMips, oldAllocatedMip, _allocatedMip, _populatedMip);
  } else {
        copyUncompressedTexGPUMem(_gpuObject, _target, oldId, _id, numMips, oldAllocatedMip, _allocatedMip, _populatedMip);
    }
    syncSampler();
});
```

File: libraries/gpu/src/gpu/Texture.cpp [hunk 6]
```diff
@@ -407,8 +408,12 @@ void Texture::setStoredMipFormat(const Element& format) {
     _storage->setFormat(format);
 }
 
-const Element& Texture::getStoredMipFormat() const {
-    return _storage->getFormat();
+Element Texture::getStoredMipFormat() const {
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.cpp]: if you make a static `Element EMPTY_ELEMENT;` here you can return it as a const reference

File: libraries/gpu/src/gpu/Texture.h [hunk 7]
```diff
@@ -285,7 +285,7 @@ class Texture : public Resource {
         Stamp bumpStamp() { return ++_stamp; }
 
         void setFormat(const Element& format) { _format = format; }
-        const Element& getFormat() const { return _format; }
+        Element getFormat() const { return _format; }
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.h]: why are you removing this const ref?

File: libraries/gpu/src/gpu/Texture.h [hunk 8]
```diff
@@ -372,7 +372,7 @@ class Texture : public Resource {
     bool isColorRenderTarget() const;
     bool isDepthStencilRenderTarget() const;
 
-    const Element& getTexelFormat() const { return _texelFormat; }
+    Element getTexelFormat() const { return _texelFormat; }
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.h]: why are you removing this const ref?

File: libraries/gpu/src/gpu/Texture.h [hunk 9]
```diff
@@ -468,7 +468,7 @@ class Texture : public Resource {
 
     // Mip storage format is constant across all mips
     void setStoredMipFormat(const Element& format);
-    const Element& getStoredMipFormat() const;
+    Element getStoredMipFormat() const;
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.h]: why are you removing this const ref?

File: libraries/gpu/src/gpu/Texture.h (hunk 7 above)
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.h]: OK, saw your earlier comment, never mind.

File: libraries/gpu-gl/src/gpu/gl41/GL41Backend.h (hunk 3 above)
Comment (samcake) [inline libraries/gpu-gl/src/gpu/gl41/GL41Backend.h]: Yes it is fine (the same way we use texUnit 34 and above for resource Buffer bound as texBUffer already in 4.1.
4.1 supports at minimum 80 binding points for textures.
THe number of textures used by one shader stage during a drawcall is more limited (it's 16) but that s a different limit.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 41f0b1682ed2a7fdb228208a3a7cfafe57cdabd5 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-41f0b16.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-41f0b16.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: dc50db7ee5f0fa59d68f4604b96af6dcb31416a3 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-dc50db7.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-dc50db7.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0ecf599267f76c0a71dfea3bccd8bfbc22942b32 available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-0ecf599.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-0ecf599.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 48eb30936eb37fe9c27f0681317325bb095e28bc available at: https://gist.github.com/edcf389327d5cdb61a6096f1c2212cf5<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-48eb309.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10313-48eb309.dmg

**Comment by Atlante45** [comment]
Tested on mac, Mips rez in and out as expected, QA PASS.

**Comment by samcake** [comment]
Tested mac/amd and nv QA passed
